File: BluetoothLightLib/src/main/java/com/telink/util/NumberUtils.java
/*
 * Copyright (C) 2015 The Telink Bluetooth Light Project
 *
 */
package com.telink.util;

public final class NumberUtils {

    private NumberUtils() {
    }

    static public int byteToInt(byte s, int bitStartPosition, int bitEndPosition) {
        int bit = bitEndPosition - bitStartPosition + 1;
        int maxValue = 1 << bit;
        int result = 0;

        for (int i = bitEndPosition, j = bit; i > bitStartPosition; i--, j--) {
            result += (s >> i & 0x01) << j;
        }

        return result & maxValue;
    }

    static public long bytesToLong(byte[] s, int start, int length) {
        int end = start + length;
        int max = length - 1;
        long result = 0;

        for (int i = start, j = max; i < end; i++, j--) {
            result += (s[i] & 0xFF) << (8 * j);
        }

        return result;
    }
}

File: BluetoothLightLib/src/main/java/com/telink/util/NumberUtils.kt
package com.telink.util

import kotlin.experimental.and


object NumberUtils {

    fun byteToInt(s: Byte, bitStartPosition: Int, bitEndPosition: Int): Int {
        val bit = bitEndPosition - bitStartPosition + 1
        val maxValue = 1 shl bit
        var result = 0

        var i = bitEndPosition
        var j = bit
        while (i > bitStartPosition) {
            var temp = s.toInt() shr i
            result += temp and 0x01 shl j
            i--
            j--
        }

        return result and maxValue
    }

    fun bytesToLong(s: ByteArray, start: Int, length: Int): Long {
        val end = start + length
        val max = length - 1
        var result = 0L

        var i = start
        var j = max
        while (i < end) {
            var temp = (s[i] and 0xFF.toByte()).toInt()
            result += ( temp shl 8 * j).toLong()
            i++
            j--
        }

        return result
    }
}
